README.md
#  Blog Time

A plugin for WordPress that displays the time according to your blog via an admin toolbar widget, a sidebar widget, and/or a template tag.

This plugin is available in the WordPress Plugin Directory: https://wordpress.org/plugins/blog-time/


## Installation

1. Install via the built-in WordPress plugin installer. Or install the plugin code inside the plugins directory for your site (typically `/wp-content/plugins/`).
2. Activate the plugin through the 'Plugins' admin menu in WordPress
3. Optionally customize the time format used for displaying the time via the "Blog Time Format" setting found in Settings -> General.
4. Optionally use the 'Blog Time' widget or the template tag `c2c_blog_time()` in a theme template file to display the blog's time at the time of the page's rendering.


## Additional Documentation

See [readme.txt](https://github.com/coffee2code/blog-time/blob/master/readme.txt) for additional usage information. See [CHANGELOG.md](CHANGELOG.md) for the list of changes for each release.


## Support

Commercial support and custom development are not presently available. You can raise an [issue](https://github.com/coffee2code/blog-time/issues) on GitHub or post in the [plugin's support forum on WordPress.org](https://wordpress.org/support/plugin/blog-time/).

If the plugin has been of benefit to you, how about [submitting a review](https://wordpress.org/support/plugin/blog-time/reviews/) for it in the WordPress Plugin Directory or considering a [donation](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=6ARCFJ9TX3522)?


## License

This plugin is free software; you can redistribute it and/or modify it under the terms of the [GNU General Public License](https://www.gnu.org/licenses/gpl-2.0.html) as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

CHANGELOG.md
# Changelog

## 4.0.1 _(2021-07-29)_
* Fix: Ensure inline script is only output once per pageload to prevent issue when blog time appears on page multiple times, which resulted in all acting as if static
* New: Add `reset()` to restore memoized class variables to default values
* Change: Note compatibility through WP 5.8+

## 4.0 _(2021-06-19)_

### Highlights:

This recommended release introduces a setting for configuring the blog time format, adds support for the 'T' timezone format token, updates the bundled Moment.js library, improves documentation, restructures unit test files, notes compatibility through 5.7+, and incorporates numerous behind-the-scenes tweaks.

### Details:

* New: Add setting for configuring time format
    * New: Add "Blog Time Format" setting to the "General Settings" page.
    * New: Add link to setting from plugin's action links
    * New: Show default time format when setting is blank
    * New: Show inline notice below setting when time format is filtered, and indicate that it takes precedence over setting
    * New: Add `initialize_setting()`, `allowed_options()`, `display_option()`, `plugin_action_links()`, and `is_wp_55_or_later()`
    * New: Add new screenshot
* Change: Use default time format if parameter or filter attempts to configure an empty string or non-string value
* Fix: Add support for the 'T' timezone format character to the dynamic clock (support for which was removed from Moment.js awhile ago)
* Change: Update bundled Moment.js to v2.29.1
    * 2.29.1: https://gist.github.com/marwahaha/cc478ba01a1292ab4bd4e861d164d99b
    * 2.29.0: https://gist.github.com/marwahaha/b0111718641a6461800066549957ec14
    * 2.28.0: https://gist.github.com/marwahaha/028fd6c2b2470b2804857cfd63c0e94f
    * 2.27.0: https://gist.github.com/marwahaha/5100c9c2f42019067b1f6cefc333daa7
* Removed: Dropped support for long-deprecated `'blog_time_format'` filter. Use `'c2c_blog_time_format'` instead.
* Change: Switch to use of `wp_add_inline_script()` instead of `wp_localize_script()`
* Change: Add optional `$exit` arg to `report_time()` to allow not exiting after outputting the time
* Change: Improve some inline documentation
* Change: Improve documentation and formatting in readme.txt
* Change: Note compatibility through WP 5.7+
* Change: Update URLs to PHP documentation for datetime formatting
* Change: Update copyright date (2021)
* Unit tests:
    * Change: Restructure unit test directories and files into `tests/` top-level directory
        * Change: Move `bin/` into `tests/`
        * Change: Move `tests/bootstrap.php` into `tests/phpunit/`
        * Change: In bootstrap, store path to plugin file constant so its value can be used within that file and in test file
        * Change: Move `tests/*.php` into `tests/phpunit/tests/`
        * Change: Remove 'test-' prefix from unit test files
        * Change: Rename `phpunit.xml` to `phpunit.xml.dist` per best practices
    * New: Add tests for `enqueue_js()`, `report_time()`, `admin_bar_menu()`
* New: Add a few more possible TODO items

## 3.6.2 _(2020-06-11)_

### Highlights:

This minor release updates the Moment.js library, adds a TODO.md file, updates a few URLs to be HTTPS, expands unit testing, and notes compatibility through WP 5.4+.

### Details:

* Change: Update Moment.js to v2.26.0
    * 2.26.0: https://gist.github.com/marwahaha/0725c40740560854a849b096ea7b7590
    * 2.25.3: https://github.com/moment/moment/blob/develop/CHANGELOG.md#2253
    * 2.25.2: https://github.com/moment/moment/blob/develop/CHANGELOG.md#2252
    * 2.25.1: https://github.com/moment/moment/blob/develop/CHANGELOG.md#2251
    * 2.25.0: https://gist.github.com/ichernev/6148e64df2427e455b10ce6a18de1a65
* Change: Remove `is_wp_login()` since it is no longer necessary
* Change: Remove redundant check in `enqueue_js()` that is already performed in `show_in_toolbar_for_user()`
* New: Add TODO.md and move existing TODO list from top of main plugin file into it
* Change: Note compatibility through WP 5.4+
* Change: Update links to coffee2code.com to be HTTPS
* Change: Add an FAQ and tweak docs in readme.txt
* Unit tests:
    * New: Add tests for `display_time()`
    * Change: Add more tests for `get_time_format()`
    * Change: Use HTTPS for link to WP SVN repository in bin script for configuring unit tests (and delete commented-out code)
    * Change: Remove unnecessary unregistering of hooks in `tearDown()`

## 3.6.1 _(2019-12-01)_
* Change: Update unit test install script and bootstrap to use latest WP unit test repo
* Change: Note compatibility through WP 5.3+
* Change: Update copyright date (2020)

## 3.6 _(2019-04-02)_
* Change: Use minimized version of Moment.js library to reduce resource usage
* Change: Update Moment.js to v2.24.0
* Change: Change handle for enqueuing Moment.js to allow only one copy of the library being enqueued now that WP packages the library
* Change: Remove `set_js_ajaxurl()` and localize `ajaxurl` alongside other variables instead of outputting it directly
* Change: Initialize plugin on `plugins_loaded` action instead of on load
* Change: Merge `do_init()` into `init()`
* Change: Update widget framework to 013
    * Add `get_config()` as a getter for config array
* Change: Update widget to 008
    * Update to use v013 of the widget framework
* Change: Cast return value of `c2c_blog_time_toolbar_widget_for_user` and `c2c_blog_time_active_clock` filters as boolean
* Change: Ensure widget markup uses double-quotes rather than single-quotes for class attribute values
* Unit tests:
    * Add unit test for `add_widget()`
    * Add unit tests for `show_in_toolbar_for_user()`
    * Add unit test for `c2c_blog_time_active_clock` filter
    * Add unit test for `c2c_blog_time_toolbar_widget_for_user` filter
    * Add unit tests for hooking of various actions
* New: Add CHANGELOG.md file and move all but most recent changelog entries into it
* New: Add inline documentation for hooks

tests/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Blog_Time
 */

ini_set( 'display_errors', 'on' );
error_reporting( E_ALL );

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/tests/phpunit/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '/blog-time.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/tests/phpunit/includes/bootstrap.php';


tests/phpunit/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Blog_Time
 */

define( 'BLOG_TIME_PLUGIN_FILE', dirname( __FILE__, 3 ) . '/blog-time.php' );

ini_set( 'display_errors', 'on' );
error_reporting( E_ALL );

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/tests/phpunit/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require BLOG_TIME_PLUGIN_FILE;
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/tests/phpunit/includes/bootstrap.php';


js/jqClock.js
/* Sets time in clock div and calls itself every second */
/**
 * Clock plugin
 * Copyright (c) 2010 John R D'Orazio (donjohn.fmmi@gmail.com)
 * Dual licensed under the MIT and GPL licenses:
 * http://www.opensource.org/licenses/mit-license.php
 * http://www.gnu.org/licenses/gpl.html
 *
 * Turns a jQuery dom element into a dynamic clock
 *
 * @timestamp defaults to clients current time
 *   $("#mydiv").clock();
 *   >> will turn div into clock using client computer's current time
 * @timestamp server-side example:
 *   Say we have a hidden input with id='timestmp' the value of which is determined server-side with server's current time
 *   $("#mydiv").clock({"timestamp":$("#timestmp").val()});
 *   >> will turn div into clock passing in server's current time as retrieved from hidden input
 *
 * @format defaults to 12 hour format,
 *   or if langSet is indicated defaults to most appropriate format for that langSet
 *   $("#mydiv").clock(); >> will have 12 hour format
 *   $("#mydiv").clock({"langSet":"it"}); >> will have 24 hour format
 *   $("#mydiv").clock({"langSet":"en"}); >> will have 12 hour format
 *   $("#mydiv").clock({"langSet":"en","format":"24"}); >> will have military style 24 hour format
 *   $("#mydiv").clock({"calendar":true}); >> will include the date with the time, and will update the date at midnight
 *
 */

(function($, undefined) {

$.clock = { version: "2.0.1", locale: {} }

t = new Array();

$.fn.clock = function(options) {
  var locale = {
    "it":{
      "weekdays":["Domenica","Lunedì","Martedì","Mercoledì","Giovedì","Venerdì","Sabato"],
      "months":["Gennaio","Febbraio","Marzo","Aprile","Maggio","Giugno","Luglio","Agosto","Settembre","Ottobre","Novembre","Dicembre"]
    },
    "en":{
      "weekdays":["Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"],
      "months":["January","February","March","April","May","June","July","August","September","October","November","December"]
    },
    "es":{
      "weekdays":["Domingo", "Lunes", "Martes", "Miércoles", "Jueves", "Viernes", "Sábado"],
      "months":["Enero", "Febrero", "Marzo", "Abril", "May", "junio", "Julio", "Agosto", "Septiembre", "Octubre", "Noviembre", "Diciembre"]
    },
    "de":{
      "weekdays":["Sonntag", "Montag", "Dienstag", "Mittwoch", "Donnerstag", "Freitag", "Samstag"],
      "months":["Januar", "Februar", "März", "April", "könnte", "Juni", "Juli", "August", "September", "Oktober", "November", "Dezember"]
    },
    "fr":{
      "weekdays":["Dimanche", "Lundi", "Mardi", "Mercredi", "Jeudi", "Vendredi", "Samedi"],
      "months":["Janvier", "Février", "Mars", "Avril", "May", "Juin", "Juillet", "Août", "Septembre", "Octobre", "Novembre", "Décembre"]
    },
    "ru":{
      "weekdays":["Воскресенье", "Понедельник", "Вторник", "Среда", "Четверг", "Пятница", "Суббота"],
      "months":["Январь", "Февраль", "Март", "Апрель", "Май", "Июнь", "Июль", "Август", "Сентябрь", "Октябрь", "Ноябрь", "Декабрь"]
    }
  }

  return this.each(function(){
    $.extend(locale,$.clock.locale);
    options = options || {};
    options.timestamp = options.timestamp || "systime";
    systimestamp = new Date();
    systimestamp = systimestamp.getTime();
    options.sysdiff = 0;
    if(options.timestamp!="systime"){
      mytimestamp = new Date(options.timestamp);
      options.sysdiff = options.timestamp - systimestamp;
    }
    options.langSet = options.langSet || "en";
    options.format = options.format || ((options.langSet!="en") ? "24" : "12");
    options.calendar = options.calendar || "true";

    if (!$(this).hasClass("jqclock")){$(this).addClass("jqclock");}

    var addleadingzero = function(i){
      if (i<10){i="0" + i;}
      return i;
    },
    updateClock = function(el,myoptions) {
      var el_id = $(el).attr("id");
      if(myoptions=="destroy"){ clearTimeout(t[el_id]); }
      else {
        mytimestamp = new Date();
        mytimestamp = mytimestamp.getTime();
        mytimestamp = mytimestamp + myoptions.sysdiff;
        mytimestamp = new Date(mytimestamp);
        var h=mytimestamp.getHours(),
        m=mytimestamp.getMinutes(),
        s=mytimestamp.getSeconds(),
        dy=mytimestamp.getDay(),
        dt=mytimestamp.getDate(),
        mo=mytimestamp.getMonth(),
        y=mytimestamp.getFullYear(),
        ap="",
        calend="";

        if(myoptions.format=="12"){
          ap=" AM";
          if (h > 11) { ap = " PM"; }
          if (h > 12) { h = h - 12; }
          if (h == 0) { h = 12; }
        }

        // add a zero in front of numbers 0-9
        h=addleadingzero(h);
        m=addleadingzero(m);
        s=addleadingzero(s);

        if(myoptions.calendar!="false") {
          if (myoptions.langSet=="en") {
            calend = "<span class='clockdate'>"+locale[myoptions.langSet].weekdays[dy]+', '+locale[myoptions.langSet].months[mo]+' '+dt+', '+y+"</span>";
          }
          else {
            calend = "<span class='clockdate'>"+locale[myoptions.langSet].weekdays[dy]+', '+dt+' '+locale[myoptions.langSet].months[mo]+' '+y+"</span>";
          }
        }
        $(el).html(calend+"<span class='clocktime'>"+h+":"+m+":"+s+ap+"</span>");
        t[el_id] = setTimeout(function() { updateClock( $(el),myoptions ) }, 1000);
      }

    }

    updateClock($(this),options);
  });
}

  return this;

})(jQuery);


js/blog-time.js
if (jQuery) {
	jQuery(document).ready(function($) {

		var c2c_blog_time_update_clock = function(c) {
			// If time format is not embedded in markup (as the case for a widget), then
			// use configured value.
			var embedded_time_format = $(c).data('time-format');
			var time_format = embedded_time_format ? embedded_time_format : c2c_BlogTime.time_format;
			var utc_offset = c2c_BlogTime.utc_offset;
			var now = moment().utcOffset(utc_offset).format(time_format);
			$(c).find('.c2c-blog-time-widget-display').html(now); //+":"+s

			// Determine update interval.
			if ( time_format.indexOf('s') > -1 ) {
				// Seconds are being displayed, so must update every second.
				interval_seconds = 1;
			} else if ( time_format.indexOf('m') > -1 ) {
				// Minutes are being displayed, so update every 10 seconds.
				interval_seconds = 10;
			} else {
				// Any longer time period is fine with an update once a minute.
				// While the interval could be made even longer, when the page is
				// loaded on the cusp of a time/date change, we don't want the
				// time display to be wrong for too long.
				interval_seconds = 60;
			}
			// Update timestamp in the determined intervals.
			setTimeout(function() { c2c_blog_time_update_clock( $(c) ) }, interval_seconds * 1000);
		}

		$.each($('.c2c-blog-time-widget'), function(){
			var dynamic_clock = $(this).find('.c2c-blog-time-dynamic');
			var display = $(this).find('.c2c-blog-time-widget-display');
			if (dynamic_clock.length > 0) { // If c2c-blog-time-dynamic is present, clock should be dynamic
				// Set a CSS class for container to denote dynamic clock.
				$(this).parent().addClass('c2c-blog-time-dynamic');
				display.find('a')
					.attr('title', '')
					.click(function() { return false; });
					c2c_blog_time_update_clock($(this));
			}
			else {
				display.click(function() {
					$.get(c2c_BlogTime.ajaxurl, {action: 'report_time'}, function(data) {
						display.find('a').html(data);
					});
					return false;
				});
			}
		});
	});
}


blog-time.php
<?php
/**
 * Plugin Name: Blog Time
 * Version:     4.0.1
 * Plugin URI:  https://coffee2code.com/wp-plugins/blog-time/
 * Author:      Scott Reilly
 * Author URI:  https://coffee2code.com/
 * Text Domain: blog-time
 * License:     GPLv2 or later
 * License URI: https://www.gnu.org/licenses/gpl-2.0.html
 * Description: Display the time according to your blog via admin toolbar widget, a sidebar widget, and/or template tag.
 *
 * Compatible with WordPress 4.6 through 5.8+.
 *
 * =>> Read the accompanying readme.txt file for instructions and documentation.
 * =>> Also, visit the plugin's homepage for additional information and updates.
 * =>> Or visit: https://wordpress.org/plugins/blog-time/
 *
 * @package Blog_Time
 * @author  Scott Reilly
 * @version 4.0.1
 */

/*
	Copyright (c) 2009-2021 by Scott Reilly (aka coffee2code)

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

defined( 'ABSPATH' ) or die();

require_once( dirname( __FILE__ ) . DIRECTORY_SEPARATOR . 'blog-time.widget.php' );

if ( ! class_exists( 'c2c_BlogTime' ) ) :

class c2c_BlogTime {

	/**
	 * The name used for the plugin's setting.
	 *
	 * @ccess private
	 * @since 4.0
	 * @var string
	 */
	private static $setting_name = 'c2c_blog_time';

	/**
	 * Internally stored configuration settings.
	 * @var array
	 */
	private static $config = array();

	/**
	 * Internally used flag to indicate if inline JS script has been added.
	 *
	 * @access private
	 * @since 4.0.1
	 * @var bool
	 */
	private static $added_inline = false;

	/**
	 * Returns version of the plugin.
	 *
	 * @since 3.0
	 */
	public static function version() {
		return '4.0.1';
	}

	/**
	 * Initialization.
	 *
	 */
	public static function init() {
		// Load textdomain.
		load_plugin_textdomain( 'blog-time' );

		self::reset();

		// Register hooks.
		add_action( 'admin_init',                 array( __CLASS__, 'initialize_setting' ) );
		add_action( 'admin_bar_menu',             array( __CLASS__, 'admin_bar_menu' ), 500 );
		add_action( 'admin_enqueue_scripts',      array( __CLASS__, 'enqueue_js' ) );
		add_action( 'wp_enqueue_scripts',         array( __CLASS__, 'enqueue_js' ) );
		add_action( 'wp_ajax_report_time',        array( __CLASS__, 'report_time' ) );
		add_action( 'wp_ajax_nopriv_report_time', array( __CLASS__, 'report_time' ) );
	}

	/**
	 * Resets the class.
	 *
	 * Primarily restores memoization variables to default values.
	 *
	 * @since 4.0.1
	 */
	public static function reset() {
		self::$config = array(
			'time_format' => __( 'g:i A', 'blog-time' )
		);

		self::$added_inline = false;
	}

	/**
	 * Determines if the running WordPress is WP 5.5 or later.
	 *
	 * @since 4.0
	 *
	 * @return bool True if WP is 5.5 or later, else false.
	 */
	public static function is_wp_55_or_later() {
		return version_compare( $GLOBALS['wp_version'], '5.5', '>=' );
	}

	/**
	 * Adds a 'Settings' link to the plugin action links.
	 *
	 * @since 4.0
	 *
	 * @param string[] $action_links An array of plugin action links.
	 * @return array
	 */
	public static function plugin_action_links( $action_links ) {
		$settings_link = sprintf(
			'<a href="%s">%s</a>',
			esc_url( admin_url( 'options-general.php' ) . '#c2c_blog_time' ),
			__( 'Settings', 'blog-time' )
		);
		array_unshift( $action_links, $settings_link );

		return $action_links;
	}

	/**
	 * Initializes setting.
	 *
	 * @since 4.0
	 */
	public static function initialize_setting() {
		if ( ! current_user_can( 'manage_options' ) ) {
			return;
		}

		register_setting( 'general', self::$setting_name );

		add_filter(
			self::is_wp_55_or_later() ? 'allowed_options' : 'whitelist_options',
			array( __CLASS__, 'allowed_options' )
		);

		add_settings_field(
			self::$setting_name,
			__( 'Blog Time Format', 'blog-time' ),
			array( __CLASS__, 'display_option' ),
			'general'
		);

		// Add link to settings page from the plugin's action links on plugin page.
		add_filter( 'plugin_action_links_blog-time/blog-time.php', array( __CLASS__, 'plugin_action_links' ) );
	}

	/**
	 * Allows the plugin's option(s).
	 *
	 * @since 4.0
	 *
	 * @param array $options Array of options.
	 * @return array The amended options array.
	 */
	public static function allowed_options( $options ) {
		$added = array( self::$setting_name => array( self::$setting_name ) );

		return self::is_wp_55_or_later()
			? add_allowed_options( $added, $options )
			: add_option_whitelist( $added, $options );
	}

	/**
	 * Outputs markup for the plugin setting on the Reading Settings page.
	 *
	 * @since 4.0
	 *
	 * @param array $args Arguments.
	 */
	public static function display_option( $args = array() ) {
		printf(
			'<input name="%s" type="text" id="%s" value="%s" class="short-text">' . "\n" . '<p class="description">%s</p>' . "\n" . '<p class="blog-time-info">%s</p>' . "\n",
			esc_attr( self::$setting_name ),
			esc_attr( self::$setting_name ),
			esc_attr( self::get_time_format( '', 'raw' ) ),
			sprintf(
				/* translators: %s: URL to PHP documentation for data and time formatting. */
				__( 'Used by the <strong>Blog Time</strong> plugin. See <a href="%s">Documentation on date and time formatting</a> for formatting syntax.', 'blog-time' ),
				'https://www.php.net/manual/en/datetime.format.php'
			),
			sprintf(
				/* translators: %s: Default blog time format. */
				__( 'Default (used when setting is blank): %s', 'blog-time' ),
				'<code>' . self::$config['time_format'] . '</code>'
			)
		);

		// Provide a warning notice when setting value is being overridden via filter.
		$time_format =self::get_time_format();
		if ( self::get_time_format( '', 'nofilter' ) !== $time_format ) {
			// Note: If a filter callback sets time format to the default (or to the setting value if one is set),
			// then this warning will not appear. That's fine since the interface conveys the time format in use,
			// just not technically accurate by implying the default/setting value is being used. If the default or
			// setting is changed, the notice will then appear, which might be surprising since there was no mention
			// of a filter being used just prior.
			printf(
				'<p class="%s blog-time-info notice notice-warning">%s</p>' . "\n",
				esc_attr(self::$setting_name ),
				sprintf(
					/* translators: 1: Filter name, 2: Filtered time format. */
					__( 'The blog time format is currently configured via the %1$s filter, which takes precedence over this setting. The filtered blog time format value is %2$s (which may differ, depending on the logic used in the filter callback).', 'blog-time' ),
					"'c2c_blog_time_format'",
					'<code>' . $time_format . '</code>'
				)
			);
		}
	}

	/**
	 * Is the blog time admin toolbar widget enabled for the specified user?
	 *
	 * TODO: This is mostly a placeholder for future functionality whereby
	 * the admin toolbar widget is controlled on a per-user basis via a
	 * user option.
	 *
	 * @since 3.0
	 *
	 * @return bool True if enabled, false if not.
	 */
	public static function show_in_toolbar_for_user() {
		/**
		 * Filters if the blog time admin toolbar widget should be enabled for user.
		 *
		 * @since 3.0
		 *
		 * @param bool $enabled_for_user Is the blog time admin toolbar widget
		 *                               enabled for user? Default true.
		 */
		return is_admin_bar_showing() ? (bool) apply_filters( 'c2c_blog_time_toolbar_widget_for_user', true ) : false;
	}

	/**
	 * Enqueues JavaScript.
	 *
	 * @since 2.0
	 *
	 * @param bool $force Optional. Should scripts get enqueued regardless of
	 *                    admin toolbar check? (e.g. when widget is displayed)
	 */
	public static function enqueue_js( $force = false ) {
		if ( ! $force && ! self::show_in_toolbar_for_user() ) {
			return;
		}

		wp_enqueue_style( __CLASS__, plugins_url( 'css/blog-time.css', __FILE__ ), array(), self::version() );

		wp_enqueue_script( 'jquery' );
		wp_enqueue_script( 'moment', plugins_url( 'js/moment.min.js', __FILE__ ), array(), '2.29.1', true );
		wp_enqueue_script( __CLASS__, plugins_url( 'js/blog-time.js', __FILE__ ), array( 'jquery', 'moment' ), self::version(), true );

		if ( ! self::$added_inline ) {
			wp_add_inline_script( __CLASS__, 'const ' . __CLASS__ . ' = ' . json_encode( array(
				'ajaxurl'     => admin_url( 'admin-ajax.php' ),
				'time_format' => self::get_time_format( '', 'momentjs' ),
				'utc_offset'  => self::display_time( 'O', 'utc-offset' ),
			) ), 'before' );
			self::$added_inline = true;
		}
	}

	/**
	 * Adds time to admin toolbar.
	 *
	 * @since 3.0
	 */
	public static function admin_bar_menu() {
		global $wp_admin_bar;

		$wp_admin_bar->add_menu( array(
			'id'     => 'c2c-blog-time',
			'parent' => 'top-secondary',
			'title'  => self::add_widget(),
			'meta'   => array( 'class' => '', 'title' => __( 'Current blog time', 'blog-time' ) )
		) );
	}

	/**
	 * Determines the time format string for the given context.
	 *
	 * Recognized contexts (though not all have special handling):
	 * - ajax:     When time is being returned via AJAX. Functionally equivalent
	 *             to 'default'.
	 * - default:  The default context. Returns the true blog time format with
	 *             everything (default value, parameter, setting, filter) taken
	 *             into account.
	 * - momentjs: When time is to be ultimately used by MomentJS. Functionally
	 *             equivalent to 'default', but with format tokens converted from
	 *             PHP-syntax to MomentJS-syntax.
	 * - nofilter: Takes default value, parameter, and setting into account, but
	 *             does not pass the blog time format through any filters.
	 * - raw:      Takes parameter and setting into account, but ignores default
	 *             value and filter. If parameter is set, then returns that value.
	 *             Else returns value of setting, even if empty.
	 * - template_tag: When time is being retrieved or displayed via the
	 *             template tag. Functionally equivalent to 'default'.
	 * - widget:   When time is being displayed within a widget. Functionally
	 *             equivalent to 'default'.
	 *
	 * @since 3.5
	 * @since 4.0 Add support for 'nofilter' and 'raw' contexts.
	 *
	 * @param  string $time_format Optional. The format for the time string, if
	 *                             being explicitly set. Default ''.
	 * @param  string $context.    Optional. The context for the time being
	 *                             displayed. Can be any custom value, but has
	 *                             special handling for 'momentjs' 'nofilter',.
	 *                             or 'raw'. Default 'default'.
	 * @return string The time format string.
	 */
	public static function get_time_format( $time_format = '', $context = 'default' ) {
		if ( ! $context ) {
			$context = 'default';
		}

		// If no time format has been explicitly specified, use setting value.
		if ( $time_format ) {
			$explicit = true;
		} else {
			$explicit = false;
			$time_format = get_option( self::$setting_name );
		}

		// Don't proceed any further if context is 'raw'.
		if ( 'raw' === $context ) {
			return $time_format;
		}

		// If no time format at this point, use default.
		if ( ! $time_format ) {
			$time_format = self::$config['time_format'];
		}

		// Filter blog time format unless context is 'nofilter'.
		if ( ! $explicit && 'nofilter' !== $context ) {
			/**
			 * Filters the time format string for a given context.
			 *
			 * @since 3.1
			 *
			 * @param string $time_format The format for the time string.
			 * @param string $context     The context.
			 */
			$time_format = apply_filters( 'c2c_blog_time_format', $time_format, $context );
		}

		// If no time format still, use default.
		if ( ! $time_format || ! is_string( $time_format ) ) {
			$time_format = self::$config['time_format'];
		}

		// If the context is momentjs, then convert time format to Moment's format.
		if ( 'momentjs' === $context ) {
			$time_format = self::map_php_time_format_to_momentjs( $time_format );
		}

		return $time_format;
	}

	/**
	 * Formats the current time (mysql) to the specified time format.
	 *
	 * @param  string $time_format Otional. The format for the time string, if not the default.
	 * @param  string $context.    Optional. The context for the time being displayed. Default ''.
	 * @return string The time string.
	 */
	public static function display_time( $time_format = '', $context = '' ) {
		$time_format = self::get_time_format( $time_format, $context );

		return date_i18n( $time_format, strtotime( current_time( 'mysql' ) ) );
	}

	/**
	 * Echoes the blog time and optionally exits (for use as AJAX responder).
	 *
	 * @since 4.0 Added `$exit` arg.
	 *
	 * @param bool $exit Optional. Exit after echoing time? Default true.
	 */
	public static function report_time( $exit = true ) {
		echo self::display_time( '', 'ajax' );

		if ( $exit ) {
			exit;
		}
	}

	/**
	 * Converts a PHP time format string into a Moment.js time format string.
	 *
	 * @since 3.5
	 *
	 * @param  string $format PHP time format string.
	 * @return string
	 */
	public static function map_php_time_format_to_momentjs( $format ) {
		// Preprocess format string to handle exceptions.
		$pre_mapping = array(
			'js' => '~~', // Moment doesn't have ordinal suffix as a separate token
		);
		foreach ( $pre_mapping as $pre_map => $remap ) {
			$format = str_replace( $pre_map, $remap, $format );
		}

		// Mappings for PHP time format tokens to Moment.js tokens.
		$mapping = array(
			// Day
			'd' => 'DD',
			'D' => 'ddd',
			'j' => 'D',
			'l' => 'dddd',
			'N' => 'E',
			'S' => '',
			'w' => 'd',
			'z' => 'DDD',
			// Week
			'W' => 'w',
			// Month
			'F' => 'MMMM',
			'm' => 'MM',
			'M' => 'MMM',
			'n' => 'M',
			't' => '',
			// Year
			'L' => '',
			'o' => 'gggg',
			'Y' => 'YYYY',
			'y' => 'YY',
			// Time
			'a' => 'a',
			'A' => 'A',
			'B' => 'SSS',
			'g' => 'h',
			'G' => 'H',
			'h' => 'hh',
			'H' => 'HH',
			'i' => 'mm',
			's' => 'ss',
			'u' => 'SSSSSS',
			'v' => 'SSS',
			// Timezone
			'e' => 'z',
			'I' => '',
			'O' => 'ZZ',
			'P' => 'Z',
			'T' => '[' . self::display_time( 'T' ) . ']',
			'Z' => '',
			// Full date/time
			'c' => 'YYYY-MM-DDTHH:mm:ssZ',
			'r' => 'ddd, DD MMM YYYY HH:mm:ss ZZ',
			'U' => 'x',
		);

		$parts = str_split( $format );

		if ( ! $parts ) {
			return $format;
		}

		$moment = array();
		foreach ( $parts as $token ) {
			$moment[] = isset( $mapping[ $token ] ) ? $mapping[ $token ] : $token;
		}
		$format = implode( '', $moment );

		// Finish mapping exceptions.
		$post_mapping = array(
			'~~' => 'Do',
		);
		foreach ( $post_mapping as $post_map => $remap ) {
			$format = str_replace( $post_map, $remap, $format );
		}

		return $format;
	}

	/**
	 * Returns the markup for the admin widget.
	 *
	 * @param array $args {
	 *     Optional. Configuration array. Default empty array.
	 *
	 *.    @type string.   $context Context for the widget, e.g. "widget". Default "admin-widget".
	 *     @type bool|null $dynamic Should the clock by dynamic? Default true, unless filtered via
	 *                              the 'c2c_blog_time_active_clock' filter.
	 *.    @type string.   $format  An explicit PHP time format string. Default ''.
	 * }
	 */
	public static function add_widget( $args = array() ) {
		$defaults = array(
			'context' => 'admin-widget',
			'dynamic' => null,
			'format'  => '',
		);
		$args = wp_parse_args( $args, $defaults );

		if ( is_null( $args['dynamic'] ) ) {
			/**
			 * Filters if the Javascript-powered dynamic clock introduced in v2.0 should
			 * be enabled or if instead the v1.x era behavior of a static timestamp that
			 * can be clicked to update the timestamp via AJAX should be enabled. By
			 * default the dynamic clock is enabled.
			 *
			 * @since 2.0
			 *
			 * @param bool $active_clock Is the blog time clock active? Default true.
			 */
			$is_dynamic = (bool) apply_filters( 'c2c_blog_time_active_clock', true );
		} else {
			$is_dynamic = true == $args['dynamic'];
		}

		$dynamic_class = $is_dynamic ? 'c2c-blog-time-dynamic' : '';

		$time = self::display_time( $args['format'], $args['context'] );

		// Data to encode into data attributes.
		$data_atts = '';
		if ( $args['format'] ) {
			$data_atts .= ' data-time-format="' . esc_attr( self::get_time_format( $args['format'], 'momentjs' ) ) . '"';
		}

		$out  = '<span class="c2c-blog-time-widget"' . $data_atts . '>';

		$out .= '<span class="ab-icon"></span>';

		$out .= '<span class="c2c-blog-time-widget-display ab-label ' . $dynamic_class . '">';
		$out .= sprintf(
			'<a class="ab-item" href="" title="%s">%s</a>',
			esc_attr__( 'Click to refresh blog time', 'blog-time' ),
			$time
		);
		$out .= "</span></span>\n";

		return $out;
	}

} // end c2c_BlogTime

add_action( 'plugins_loaded', array( 'c2c_BlogTime', 'init' ) );


// Template tag
if ( ! function_exists( 'c2c_blog_time' ) ) {
	/**
	 * Template tag to display the blog's time.
	 *
	 * @since 1.1
	 *
	 * @param  string $time_format PHP-style datetime format string. Uses plugin default if not specified.
	 * @param  bool   $echo        Optional. Echo the time to the page?
	 * @return string The formatted blog time.
	 */
	function c2c_blog_time( $time_format = '', $echo = true ) {
		$val = c2c_BlogTime::display_time( $time_format, 'template-tag' );
		if ( $echo ) { echo $val; }
		return $val;
	}
	add_filter( 'c2c_blog_time', 'c2c_blog_time', 10, 2 );
}

endif; // end if !class_exists()
